### ISSUE TITLE
Remove React/Webpacker commands from install steps? (all effects already included in git/yarn)

### ISSUE THREAD

**Issue by jywarren** [publiclab/plots2#9502]
@icarito and I closely reviewed the `webpacker` and `react` install steps that were recently added to different parts of our system here, in the following PRs/issues and maybe more:

- https://github.com/publiclab/plots2/issues/9392
- https://github.com/publiclab/plots2/pull/9176
- https://github.com/publiclab/plots2/issues/9382
- https://github.com/publiclab/plots2/blob/999264ebe70d750f1836a4b4dcbdb7118595639b/.github/workflows/tests.yml#L49

We believe they add a few static files which are now part of our repo in `/app/assets/...` and some modifications to our `package.json` and `yarn.lock` files which are managed by those systems, and so we think there may be no further need for these install commands -- we're removing them from production deployment in https://github.com/publiclab/plots2/pull/9501 for example.

We're wondering if we should remove this from the README as well as other places around the system since it seems like they may be redundant now that all the changes are managed in either Git or Yarn. 

Thoughts on this?

An in-depth discussion happened in https://github.com/publiclab/plots2/issues/9382 so I'm wondering how we follow up on that. 

Is it possible that the fix for that, which re-ordered steps in https://github.com/publiclab/plots2/pull/9395, was not right? And that instead we should have removed the step entirely?

I know @noi5e is a bit offline right now and there's no big rush to this. Just making sure we don't forget to think this through and make adjustments if needed. cc @RuthNjeri @cesswairimu @waridrox and others who've been involved in the conversation! Thanks!

**Comment by RuthNjeri** [comment]
Thanks, Jeff! As For the README steps in running the Web packer commands, I noticed that each time I run the command, it updates the `Package.json` file and `Yarn`, which might be confusing when pushing code that does not touch on the React Files...It's like if running bundle install would update the gem lock file, this would cause a situation where gems are updated due to the local system of a user and not the application...

**Comment by jywarren** [comment]
Yes, i think that also aligns with our understanding that all of the effects of running those commands is reflected in the package.json and yarn.lock files, so actually running them again and again is not necessary, but ALSO that now that those changes have been git tracked, even new users of the code need not run them as long as they run `yarn install` as they would already normally. Does that seem right to you?

**Comment by waridrox** [comment]
Agreed with @RuthNjeri 🙌 , running the cmds does actually update the `package.json` and `yarn.lock` files. But if `yarn install` can already take care about those dependencies then I believe the natural intuition would be to remove the cmds...

**Comment by jywarren** [comment]
OK, thanks both of you. It changes package.json and yarn.lock, but
presumably it doesn't keep adding new things we need, right? Those were
already added in the original PR by @noi5e, i believe?

In any case I'll check back in with @icarito on this and perhaps we'll move
ahead with removing them. Thanks!!

**Comment by RuthNjeri** [comment]
I was testing the yarn install command locally to see if the application will run without having to run the webpacker command...after running yarn install I needed to run bundle install to update the gems and the bin folder of the application was updated .... not sure if this is an isolated case for me...
<img width="585" alt="Screenshot 2021-04-18 at 21 08 20" src="https://user-images.githubusercontent.com/7622875/115155965-d5554b80-a08a-11eb-82fa-b2b83a3e7b8e.png">

Also, when I ran rails s, I received some webpacker errors
<img width="996" alt="Screenshot 2021-04-18 at 21 11 58" src="https://user-images.githubusercontent.com/7622875/115156060-39780f80-a08b-11eb-912d-857587cddca8.png">

I had to run  `rails webpacker:install:react `for the application to load...

Interestingly, the only file changes that happened was when I run bundle install. The yarn command and webpacker command did not create any file updates.

**Comment by jywarren** [comment]
Thanks @RuthNjeri -- but isn't `yarn install` already part of our install process? 

What was it about yarn install that caused you to need to update gems? 

I'm looking especially at `rails g webpacker:install:react` to double-check what it does:

https://github.com/reactjs/react-rails#rails-5x

https://github.com/reactjs/react-rails/blob/f8148d57a6d96b27bddfaef9f9bf26e233b27bd6/Rakefile#L26-L29

this runs `yarn upgrade` in the `react-builds` directory. @RuthNjeri is it possible that `rails g webpacker:install:react` simply runs `yarn upgrade`? or runs it in some react directory we have? 

Second, i do notice a slight mismatch between what we have, `rails g webpacker:install:react` and the docs in `react-rails`, which has no `g` -- `rails webpacker:install:react`. But not sure that's a problem.

**Comment by jywarren** [comment]
Also, in general it seems like there is evolving guidance on how to completely switch from Rails Asset Pipeline to Webpacker using this gem: https://github.com/reactjs/react-rails/issues/1052 which could be good to look at.

**Comment by RuthNjeri** [comment]
Hey Jeff, addressing some of the questions now...
> Thanks @RuthNjeri -- but isn't `yarn install` already part of our install process?
> What was it about yarn install that caused you to need to update gems?
Yes, `yarn install` is already part of our installation process. I did try rerunning yarn install again just now and the gems were not affected, so it does seem like an isolated case or a one time thing that happens. 

> this runs `yarn upgrade` in the `react-builds` directory. @RuthNjeri is it possible that simply runs `yarn upgrade`? or runs it in some react directory we have?

Yes, I think it does run a yarn upgrade, according to this [tutorial](https://medium.com/learnetto/free-tutorial-how-to-use-react-with-webpacker-and-rails-5-1-92af8e8d9d63), it also makes configuration changes and creates a React component file/folder.
<img width="782" alt="Screenshot 2021-05-16 at 19 22 30" src="https://user-images.githubusercontent.com/7622875/118404663-80eabf00-b67c-11eb-8ee0-fe96f796e78e.png">


> Second, i do notice a slight mismatch between what we have, `rails g webpacker:install:react` and the docs in `react-rails`, which has no `g` -- `rails webpacker:install:react`. But not sure that's a problem.

The`rails g webpacker:install:react` command seems to introduce new files and folders, like it is creating something from scratch
<img width="1370" alt="Screenshot 2021-05-16 at 19 34 15" src="https://user-images.githubusercontent.com/7622875/118404963-e7bca800-b67d-11eb-9d00-ef3fc05137fa.png"> 
<img width="501" alt="Screenshot 2021-05-16 at 19 30 54" src="https://user-images.githubusercontent.com/7622875/118404960-e3908a80-b67d-11eb-81eb-c51485b9ea51.png">
Come to think of it, these files and folders created are similar to the once I encountered here https://github.com/publiclab/plots2/issues/9502#issuecomment-822036027

I also ran the `rails  webpacker:install:react` without a --g, it just seems to do some update of some sort without adding any files and folders
<img width="1306" alt="Screenshot 2021-05-16 at 19 33 28" src="https://user-images.githubusercontent.com/7622875/118405042-3e29e680-b67e-11eb-96c2-ed9342b85f65.png">

I suspect that `rails g webpacker:install:react` should only be ran once when you are initially building the app and then from then on, run it without the g command ...it could also be that the g command is not needed at all...

**Comment by RuthNjeri** [comment]
> Also, in general it seems like there is evolving guidance on how to completely switch from Rails Asset Pipeline to Webpacker using this gem: [reactjs/react-rails#1052](https://github.com/reactjs/react-rails/issues/1052) which could be good to look at.

Should we look into updating our steps to reflect this? Or should we think about a total switch to the gem?

**Comment by jywarren** [comment]
"g" may refer to "generate" if it's mirroring `rake g` in Rails style. So that matches your observation!

As to the asset pipeline stuff, i wonder if it's best we make the minimal intervention now, as i think Rails 6 may have a much more fully integrated solution and is coming up in the next year anyways. 

So what should be our next step here? Perhaps, given that:

> I suspect that `rails g webpacker:install:react` should only be ran once

and that after then, we get the equivalent already using `yarn upgrade` which is already part of our build process, maybe we just remove these lines from our steps? Because the files created by `rails g webpacker:install:react` are already in our git tree.

How does this sound, @noi5e @RuthNjeri?

**Comment by RuthNjeri** [comment]
Hi @jywarren 

The next steps make sense, we could start out removing it from our steps, if that doesn't work, we can have the command without the g option... `rails webpacker:install:react`

**Comment by jywarren** [comment]
Great, thanks. @noi5e i know you have a little bandwith on this now, did you have a take on this, does it sound OK?

**Comment by jywarren** [comment]
Looking in recent logs, i see this!

```
Run bundle exec rails g webpacker:install && bundle exec rails g webpacker:install:react && bundle exec rails g react:install
Could not find generator 'webpacker:install'. Maybe you meant 'react:install', 'responders:install' or 'atwho:install'
Run `rails generate --help` for more options.
Could not find generator 'webpacker:install:react'. Maybe you meant 'react:install', 'responders:install' or 'atwho:install'
Run `rails generate --help` for more options.
       exist  app/javascript/components
   identical  app/javascript/components/.keep
```

Does that mean those two aren't even running?

**Comment by noi5e** [comment]
Phew 😅 I have a lot of catching up to do. Sorry to leave this unread for so long, I definitely appreciate all the research that everyone did here, including @RuthNjeri in [this comment](https://github.com/publiclab/plots2/issues/9502#issuecomment-841842079).

> Looking in recent logs, i see this!

@jywarren I think this error seems to be the result of a typo. Check out the [suggested commands](https://github.com/reactjs/react-rails#3-now-run-the-installers) in the `react-rails` gem:
```
$ bundle install
$ rails webpacker:install
$ rails webpacker:install:react
$ rails generate react:install
```

Yet we still have the wrong commands in `tests.yml`, for example:
https://github.com/publiclab/plots2/blob/f41f8e75c7674d5cad14b0a4afccb19df6b7c9af/.github/workflows/tests.yml#L65

I think that should be `rails webpacker:install` and `rails webpacker:install:react` (ie. remove the `g`/`generate`). It looks like only the `rails generate react:install` command has the `g`/`generate` option.

What config file does that error correspond to? That stacktrace you mention @jywarren seems to have the wrong commands according to this line: `Run bundle exec rails g webpacker:install && bundle exec rails g webpacker:install:react && bundle exec rails g react:install`

**Comment by noi5e** [comment]
> As to the asset pipeline stuff, i wonder if it's best we make the minimal intervention now, as i think Rails 6 may have a much more fully integrated solution and is coming up in the next year anyways.
> 
> So what should be our next step here? Perhaps, given that:
> 
> > I suspect that `rails g webpacker:install:react` should only be ran once
> 
> and that after then, we get the equivalent already using `yarn upgrade` which is already part of our build process, maybe we just remove these lines from our steps? Because the files created by `rails g webpacker:install:react` are already in our git tree.

I vote for going for the minimal approach for now (aka NOT totally switching over to managing ALL assets with `webpacker`, that sounds like it would be a HUGE sysadmin challenge).

Looking through this thread, and reviewing the research that everyone did, it does really seem like the commands aren't necessary, and they simply do the following things:
- install some gems and JS packages in `Gemfile` and `package.json`.
- create new folders and files

Am I missing anything? It would be nice to know for sure exactly what these commands do.

I noticed @jywarren mentioned these lines in the `react-rails` gem here:
https://github.com/reactjs/react-rails/blob/f8148d57a6d96b27bddfaef9f9bf26e233b27bd6/Rakefile#L26-L29

Am I correct to assume that this is where/how the `rails webpacker:install && bundle exec rails webpacker:install:react && bundle exec rails g react:install` commands are defined?

**Comment by noi5e** [comment]
It looks like `.gitpod.yml` has the wrong commands in it:
https://github.com/publiclab/plots2/blob/f41f8e75c7674d5cad14b0a4afccb19df6b7c9af/.gitpod.yml#L21

The first two commands should NOT have the `g`/`generate` in them.

**Comment by jywarren** [comment]
Oof, talk about a long time, that's me now. But circling back now that our summer fellowships have wrapped up! I'll start with the .gitpod.yml correction which is easy. https://github.com/publiclab/plots2/pull/10137

**Comment by jywarren** [comment]
> Am I correct to assume that this is where/how the `rails webpacker:install && bundle exec rails webpacker:install:react && bundle exec rails g react:install` commands are defined?

Yes I believe that's right, that's where I traced them too. 

OK, so I'll fix the extra `g`s in various places, and then we'll move forward on removing the commands that just install extra JS and folders. I'll do one more review of the exact commands before doing so. That way we do this in discrete pieces. Thanks all!

**Comment by jywarren** [comment]
OK, removed from GitHub Actions too: https://github.com/publiclab/plots2/pull/10138

Same as where I removed from Gitpod config: #10137

**Comment by jywarren** [comment]
OK, in #10137 i see this message in the output in GitPod (cc @noi5e):

```
    conflict  config/webpacker.yml
Overwrite /workspace/plots2/config/webpacker.yml? (enter "h" for help) [Ynaqdhm]
```

We could decide to move fwd with removing that entire line, and see what it does in GitPod?

**Comment by jywarren** [comment]
Lots more info at https://github.com/publiclab/plots2/pull/10137#issuecomment-919268106 on this now...

**Comment by jywarren** [comment]
Fixed in #10137 thankfully!!!

**Comment by jywarren** [comment]
I believe that's the last instance?
